File: apps/web-static/verify-deployment.mjs
#!/usr/bin/env node

/**
 * Static Site Deployment Verification Script
 * Verifies that the static site is properly deployed and accessible
 */

import https from 'https';
import http from 'http';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Configuration
const SITE_URL = process.env.SITE_URL || 'https://compify.com';
const TIMEOUT = 10000; // 10 seconds

// Colors for console output
const colors = {
    reset: '\x1b[0m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    cyan: '\x1b[36m'
};

function printStatus(message) {
    console.log(`${colors.green}[INFO]${colors.reset} ${message}`);
}

function printError(message) {
    console.log(`${colors.red}[ERROR]${colors.reset} ${message}`);
}

function printWarning(message) {
    console.log(`${colors.yellow}[WARN]${colors.reset} ${message}`);
}

function printHeader(message) {
    console.log(`${colors.cyan}${message}${colors.reset}`);
}

// Test pages that should be available
const testPages = [
    '/',
    '/about/',
    '/rules/',
    '/timeline/',
    '/sponsors/',
    '/faq/'
];

// Expected SEO meta tags
const expectedMetaTags = [
    'title',
    'description',
    'og:title',
    'og:description',
    'og:type',
    'og:url'
];

function makeRequest(url) {
    return new Promise((resolve, reject) => {
        const client = url.startsWith('https:') ? https : http;
        
        const req = client.get(url, { timeout: TIMEOUT }, (res) => {
            let data = '';
            
            res.on('data', (chunk) => {
                data += chunk;
            });
            
            res.on('end', () => {
                resolve({
                    statusCode: res.statusCode,
                    headers: res.headers,
                    body: data
                });
            });
        });
        
        req.on('error', reject);
        req.on('timeout', () => {
            req.destroy();
            reject(new Error('Request timeout'));
        });
    });
}

function checkSEOMetaTags(html) {
    const missingTags = [];
    
    expectedMetaTags.forEach(tag => {
        const patterns = [
            new RegExp(`<title[^>]*>.*?</title>`, 'i'),
            new RegExp(`<meta[^>]*name=["']${tag}["'][^>]*>`, 'i'),
            new RegExp(`<meta[^>]*property=["']${tag}["'][^>]*>`, 'i')
        ];
        
        const found = patterns.some(pattern => pattern.test(html));
        if (!found && tag !== 'title') {
            // Special case for title tag
            if (tag === 'title' && !/<title[^>]*>.*?<\/title>/i.test(html)) {
                missingTags.push(tag);
            }
        } else if (tag === 'title' && !/<title[^>]*>.*?<\/title>/i.test(html)) {
            missingTags.push(tag);
        } else if (!found) {
            missingTags.push(tag);
        }
    });
    
    return missingTags;
}

function checkCacheHeaders(headers) {
    const issues = [];
    
    // Check for cache control headers
    if (!headers['cache-control']) {
        issues.push('Missing Cache-Control header');
    }
    
    // Check for security headers
    const securityHeaders = [
        'x-frame-options',
        'x-content-type-options',
        'x-xss-protection'
    ];
    
    securityHeaders.forEach(header => {
        if (!headers[header]) {
            issues.push(`Missing security header: ${header}`);
        }
    });
    
    return issues;
}

async function verifyStaticSite() {
    printHeader('🔍 Verifying Static Site Deployment');
    console.log(`Testing site: ${SITE_URL}`);
    console.log('');
    
    let allTestsPassed = true;
    const results = [];
    
    for (const page of testPages) {
        const url = `${SITE_URL}${page}`;
        
        try {
            printStatus(`Testing ${page}...`);
            const response = await makeRequest(url);
            
            const result = {
                page,
                url,
                statusCode: response.statusCode,
                issues: []
            };
            
            // Check status code
            if (response.statusCode !== 200) {
                result.issues.push(`HTTP ${response.statusCode} (expected 200)`);
                allTestsPassed = false;
            }
            
            // Check content type
            const contentType = response.headers['content-type'];
            if (!contentType || !contentType.includes('text/html')) {
                result.issues.push(`Invalid content type: ${contentType}`);
                allTestsPassed = false;
            }
            
            // Check for basic HTML structure
            if (!response.body.includes('<html') || !response.body.includes('</html>')) {
                result.issues.push('Invalid HTML structure');
                allTestsPassed = false;
            }
            
            // Check SEO meta tags
            const missingSEOTags = checkSEOMetaTags(response.body);
            if (missingSEOTags.length > 0) {
                result.issues.push(`Missing SEO tags: ${missingSEOTags.join(', ')}`);
                // Don't fail for missing SEO tags, just warn
                printWarning(`${page}: Missing SEO tags: ${missingSEOTags.join(', ')}`);
            }
            
            // Check cache headers
            const cacheIssues = checkCacheHeaders(response.headers);
            if (cacheIssues.length > 0) {
                result.issues.push(...cacheIssues);
                // Don't fail for cache headers, just warn
                printWarning(`${page}: ${cacheIssues.join(', ')}`);
            }
            
            // Check for navigation
            if (!response.body.includes('nav') && !response.body.includes('navigation')) {
                result.issues.push('No navigation found');
                printWarning(`${page}: No navigation found`);
            }
            
            results.push(result);
            
            if (result.issues.length === 0) {
                printStatus(`✓ ${page} - OK`);
            } else {
                const criticalIssues = result.issues.filter(issue => 
                    issue.includes('HTTP') || issue.includes('Invalid')
                );
                if (criticalIssues.length > 0) {
                    printError(`✗ ${page} - ${criticalIssues.join(', ')}`);
                } else {
                    printStatus(`✓ ${page} - OK (with warnings)`);
                }
            }
            
        } catch (error) {
            printError(`✗ ${page} - ${error.message}`);
            results.push({
                page,
                url,
                statusCode: 0,
                issues: [error.message]
            });
            allTestsPassed = false;
        }
    }
    
    // Performance check
    console.log('');
    printHeader('⚡ Performance Check');
    
    try {
        const startTime = Date.now();
        await makeRequest(SITE_URL);
        const loadTime = Date.now() - startTime;
        
        if (loadTime < 2000) {
            printStatus(`✓ Load time: ${loadTime}ms (Good)`);
        } else if (loadTime < 5000) {
            printWarning(`⚠ Load time: ${loadTime}ms (Acceptable)`);
        } else {
            printError(`✗ Load time: ${loadTime}ms (Too slow)`);
            allTestsPassed = false;
        }
    } catch (error) {
        printError(`✗ Performance check failed: ${error.message}`);
        allTestsPassed = false;
    }
    
    // Summary
    console.log('');
    printHeader('📊 Deployment Verification Summary');
    
    const passedTests = results.filter(r => r.statusCode === 200).length;
    const totalTests = results.length;
    
    console.log(`Pages tested: ${totalTests}`);
    console.log(`Pages accessible: ${passedTests}`);
    console.log(`Success rate: ${Math.round((passedTests / totalTests) * 100)}%`);
    
    if (allTestsPassed) {
        printStatus('🎉 All critical tests passed! Static site deployment verified.');
        return true;
    } else {
        printError('❌ Some critical tests failed. Please check the issues above.');
        return false;
    }
}

// Run verification if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
    verifyStaticSite()
        .then(success => {
            process.exit(success ? 0 : 1);
        })
        .catch(error => {
            printError(`Verification failed: ${error.message}`);
            process.exit(1);
        });
}

export { verifyStaticSite };

File: apps/backend/verify-deployment.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

// Colors for console output
const (
	ColorReset  = "\033[0m"
	ColorRed    = "\033[31m"
	ColorGreen  = "\033[32m"
	ColorYellow = "\033[33m"
	ColorCyan   = "\033[36m"
)

type HealthResponse struct {
	Status    string            `json:"status"`
	Timestamp string            `json:"timestamp"`
	Version   string            `json:"version,omitempty"`
	Uptime    string            `json:"uptime,omitempty"`
	Checks    map[string]string `json:"checks,omitempty"`
}

type VerificationResult struct {
	Endpoint    string
	StatusCode  int
	ResponseTime time.Duration
	Issues      []string
	Success     bool
}

func printStatus(message string) {
	fmt.Printf("%s[INFO]%s %s\n", ColorGreen, ColorReset, message)
}

func printError(message string) {
	fmt.Printf("%s[ERROR]%s %s\n", ColorRed, ColorReset, message)
}

func printWarning(message string) {
	fmt.Printf("%s[WARN]%s %s\n", ColorYellow, ColorReset, message)
}

func printHeader(message string) {
	fmt.Printf("%s%s%s\n", ColorCyan, message, ColorReset)
}

func makeRequest(url string, timeout time.Duration) (*http.Response, time.Duration, error) {
	client := &http.Client{
		Timeout: timeout,
	}
	
	start := time.Now()
	resp, err := client.Get(url)
	duration := time.Since(start)
	
	return resp, duration, err
}

func verifyEndpoint(baseURL, endpoint string) VerificationResult {
	url := strings.TrimSuffix(baseURL, "/") + endpoint
	result := VerificationResult{
		Endpoint: endpoint,
		Issues:   []string{},
	}
	
	printStatus(fmt.Sprintf("Testing %s...", endpoint))
	
	resp, duration, err := makeRequest(url, 10*time.Second)
	result.ResponseTime = duration
	
	if err != nil {
		result.Issues = append(result.Issues, fmt.Sprintf("Request failed: %v", err))
		return result
	}
	defer resp.Body.Close()
	
	result.StatusCode = resp.StatusCode
	
	// Check status code
	expectedStatus := 200
	if endpoint == "/health" {
		expectedStatus = 200
	} else if strings.HasPrefix(endpoint, "/auth/") {
		// Auth pages might return 405 if not properly configured for GET
		// or 200 if they serve login/register forms
		if resp.StatusCode != 200 && resp.StatusCode != 405 {
			result.Issues = append(result.Issues, fmt.Sprintf("HTTP %d (expected 200 or 405)", resp.StatusCode))
		}
	} else {
		if resp.StatusCode != expectedStatus {
			result.Issues = append(result.Issues, fmt.Sprintf("HTTP %d (expected %d)", resp.StatusCode, expectedStatus))
		}
	}
	
	// Check content type
	contentType := resp.Header.Get("Content-Type")
	if endpoint == "/health" {
		if !strings.Contains(contentType, "application/json") {
			result.Issues = append(result.Issues, fmt.Sprintf("Invalid content type: %s (expected JSON)", contentType))
		}
	} else if strings.HasPrefix(endpoint, "/auth/") {
		// Auth endpoints might return different content types depending on configuration
		// 405 responses typically return text/plain, which is acceptable
		if resp.StatusCode == 200 && !strings.Contains(contentType, "text/html") {
			result.Issues = append(result.Issues, fmt.Sprintf("Invalid content type: %s (expected HTML for 200 response)", contentType))
		}
		// For 405 responses, text/plain is acceptable
	}
	
	// Read response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		result.Issues = append(result.Issues, fmt.Sprintf("Failed to read response: %v", err))
		return result
	}
	
	// Validate health endpoint response
	if endpoint == "/health" {
		var health HealthResponse
		if err := json.Unmarshal(body, &health); err != nil {
			result.Issues = append(result.Issues, fmt.Sprintf("Invalid JSON response: %v", err))
		} else {
			if health.Status != "ok" && health.Status != "healthy" {
				result.Issues = append(result.Issues, fmt.Sprintf("Unhealthy status: %s", health.Status))
			}
			if health.Timestamp == "" {
				result.Issues = append(result.Issues, "Missing timestamp in health response")
			}
		}
	}
	
	// Check security headers
	securityHeaders := map[string]string{
		"X-Content-Type-Options": "nosniff",
		"X-Frame-Options":        "", // Should be present
		"X-XSS-Protection":       "", // Should be present
	}
	
	for header, expectedValue := range securityHeaders {
		value := resp.Header.Get(header)
		if value == "" {
			printWarning(fmt.Sprintf("%s: Missing security header: %s", endpoint, header))
		} else if expectedValue != "" && value != expectedValue {
			printWarning(fmt.Sprintf("%s: Security header %s has value '%s' (expected '%s')", endpoint, header, value, expectedValue))
		}
	}
	
	result.Success = len(result.Issues) == 0
	return result
}

func verifyBackendDeployment() bool {
	backendURL := os.Getenv("BACKEND_URL")
	if backendURL == "" {
		backendURL = "http://localhost:8080"
	}
	
	printHeader("🔍 Verifying Backend Server Deployment")
	fmt.Printf("Testing server: %s\n", backendURL)
	fmt.Println()
	
	// Test endpoints
	endpoints := []string{
		"/health",
		"/auth/login",
		"/auth/register",
	}
	
	results := []VerificationResult{}
	allTestsPassed := true
	
	for _, endpoint := range endpoints {
		result := verifyEndpoint(backendURL, endpoint)
		results = append(results, result)
		
		if result.Success {
			printStatus(fmt.Sprintf("✓ %s - OK (%dms)", endpoint, result.ResponseTime.Milliseconds()))
		} else {
			printError(fmt.Sprintf("✗ %s - %s", endpoint, strings.Join(result.Issues, ", ")))
			allTestsPassed = false
		}
	}
	
	// Performance check
	fmt.Println()
	printHeader("⚡ Performance Check")
	
	healthResult := results[0] // Health endpoint
	if healthResult.ResponseTime < 1000*time.Millisecond {
		printStatus(fmt.Sprintf("✓ Health check response time: %dms (Good)", healthResult.ResponseTime.Milliseconds()))
	} else if healthResult.ResponseTime < 5000*time.Millisecond {
		printWarning(fmt.Sprintf("⚠ Health check response time: %dms (Acceptable)", healthResult.ResponseTime.Milliseconds()))
	} else {
		printError(fmt.Sprintf("✗ Health check response time: %dms (Too slow)", healthResult.ResponseTime.Milliseconds()))
		allTestsPassed = false
	}
	
	// Summary
	fmt.Println()
	printHeader("📊 Deployment Verification Summary")
	
	passedTests := 0
	for _, result := range results {
		if result.Success {
			passedTests++
		}
	}
	
	fmt.Printf("Endpoints tested: %d\n", len(results))
	fmt.Printf("Endpoints accessible: %d\n", passedTests)
	fmt.Printf("Success rate: %d%%\n", (passedTests*100)/len(results))
	
	if allTestsPassed {
		printStatus("🎉 All critical tests passed! Backend deployment verified.")
		return true
	} else {
		printError("❌ Some critical tests failed. Please check the issues above.")
		return false
	}
}

func main() {
	success := verifyBackendDeployment()
	if success {
		os.Exit(0)
	} else {
		os.Exit(1)
	}
}